File: java/src/main/java/com/ytta/v0/Strategy.java
package com.ytta.v0;

import com.ytta.v0.Types.ActionKind;
import com.ytta.v0.Types.Side;
import com.ytta.v0.Types.StrategyAction;
import com.ytta.v0.Types.Tick;

public final class Strategy {
    private long nextClOrdId = 1;
    private Long cancelTarget = null;

    public StrategyAction onTick(Tick tick) {
        return switch ((int) tick.seq()) {
            case 1 -> {
                long id = nextClOrdId++;
                yield new StrategyAction(
                        ActionKind.NEW_ORDER, tick.tsNs(), id, Side.BUY, 100, 5, "resting_bid");
            }
            case 2 -> {
                long id = nextClOrdId++;
                yield new StrategyAction(
                        ActionKind.NEW_ORDER, tick.tsNs(), id, Side.SELL, 100, 5, "cross_self");
            }
            case 3 -> {
                long id = nextClOrdId++;
                cancelTarget = id;
                yield new StrategyAction(
                        ActionKind.NEW_ORDER, tick.tsNs(), id, Side.BUY, 99, 1, "resting_bid");
            }
            case 4 -> new StrategyAction(
                    ActionKind.CANCEL,
                    tick.tsNs(),
                    cancelTarget == null ? 0 : cancelTarget,
                    Side.BUY,
                    0,
                    0,
                    "done");
            default -> new StrategyAction(
                    ActionKind.NOOP, tick.tsNs(), 0, Side.BUY, 0, 0, "flat");
        };
    }
}

File: go/internal/v0/strategy.go
package v0

type Strategy struct {
	nextClOrdID  uint64
	cancelTarget uint64
	hasCancel    bool
}

func NewStrategy() *Strategy {
	return &Strategy{nextClOrdID: 1}
}

func (s *Strategy) OnTick(tick Tick) StrategyAction {
	a := StrategyAction{TsNs: tick.TsNs}
	switch tick.Seq {
	case 1:
		a.Kind = ActionNewOrder
		a.ClOrdID = s.nextClOrdID
		s.nextClOrdID++
		a.Side = SideBuy
		a.Price = 100
		a.Qty = 5
		a.Reason = "resting_bid"
		return a
	case 2:
		a.Kind = ActionNewOrder
		a.ClOrdID = s.nextClOrdID
		s.nextClOrdID++
		a.Side = SideSell
		a.Price = 100
		a.Qty = 5
		a.Reason = "cross_self"
		return a
	case 3:
		a.Kind = ActionNewOrder
		a.ClOrdID = s.nextClOrdID
		s.nextClOrdID++
		a.Side = SideBuy
		a.Price = 99
		a.Qty = 1
		a.Reason = "resting_bid"
		s.cancelTarget = a.ClOrdID
		s.hasCancel = true
		return a
	case 4:
		a.Kind = ActionCancel
		if s.hasCancel {
			a.ClOrdID = s.cancelTarget
		}
		a.Reason = "done"
		return a
	default:
		a.Kind = ActionNoop
		a.Reason = "flat"
		return a
	}
}
